### PR TITLE
Shuttle manipulator

### PR DESCRIPTION
**Pull request by coiax**
#16761 is in a stable but crude state.

This PR is to add a more streamlined interface for admins to select shuttle templates instead of just crudely hitting verbs.
- Removes shuttle verbs.
- Gives features to shuttle manipulator at Centcom.
- Look at a list of shuttle templates that are appropriate for the one you want to replace.
- Also monitor all the shuttles I guess.
- It allows hotswapping of shuttles while they're recalling or being summoned.
- Also some GC changes to allow us to hard delete items like greentext or mobile docking ports.

:cl:
rscdel: Emagging the emergency shuttle computer with less than ten seconds until launch no longer gives additional time.
rscadd: The emergency shuttle computer now reads the ID in your ID slot, rather than your hand.
/:cl:
- Added a new special statue
- All shuttle templates are now "timid" and will not register themselves with the shuttles subsystem on load
- Barshuttle now has a Barmaiden and anti greytide tables
- All items that return QDEL_HINT_LETMELIVE must also take the force parameter, and call super and return it if force = TRUE.
- Shuttles now give visible messages when road killing mobs
### Interface

image
image
image

### CODE REVIEW

**Comment by Dorsisdwarf** [comment]
Mother of chrome, the comits

**Comment by TrustyGun** [comment]
untitled

**Comment by coiax** [comment]
73 of these commits are in the shuttles PR, which I expect to be merged within the week. And it's been actively worked on, I don't know why so many commits are bad.

**Comment by TrustyGun** [comment]
its not bad, i just found it funny how i had to zoom out 3 times for all of them to show

**Comment by Durkel** [comment]
I guess you could say coiax is commited to the pr.

**Comment by PKPenguin321** [comment]
coiax do you not know that you can branch off of master?

**Comment by coiax** [comment]
I'm making a PR that requires another PR to be merged. I can't branch off on master.

**Comment by Bawhoppen** [comment]
coiax why you weren't supposed to make a pr yet did you even listen to me 
reeeee

**Comment by coiax** [comment]
@Bawhoppen, this isn't your player made shuttles, this is just an extension of my shuttle stuff, which I'm doing in a new branch so the existing stuff gets merged

**Comment by Bawhoppen** [comment]
I assumed it was part of it is what I meant

**Comment by neersighted** [comment]
you are a brave man. ping me when you're ready for review

**Comment by coiax** [comment]
@neersighted ping ping review

File: code/modules/shuttle/manipulator.dm [hunk 1]
```diff
@@ -9,5 +9,214 @@
...
+				"templates" = list())
+
+		var/list/L = list()
+		L["name"] = S.name
+		L["shuttle_id"] = S.shuttle_id
+		L["port_id"] = S.port_id
+		L["description"] = S.description
+		L["admin_notes"] = S.admin_notes
+
+		if(selected && (selected.shuttle_id == S.shuttle_id))
```
Comment (neersighted) [inline code/modules/shuttle/manipulator.dm]: I'd prefer that this be a flag of some sort... Maybe have selected be the name? Sending data down the wire twice when it can be derived is wasteful.

(a computed variable could easily be used)

File: code/modules/shuttle/manipulator.dm [hunk 2]
```diff
@@ -9,5 +9,214 @@
...
+			data["selected"] = L
+
+		templates[S.port_id]["templates"] += list(L)
+
+	data["templates_tabs"] = sortList(data["templates_tabs"])
```
Comment (neersighted) [inline code/modules/shuttle/manipulator.dm]: It's clearer to use a temporary var first and then assign to data, but it's up to your preference

File: code/modules/shuttle/manipulator.dm [hunk 3]
```diff
@@ -9,5 +9,214 @@
...
+	// PREVIEW button to load
+	// CANCEL button to (unload if loaded) and return back
+	// LOAD button
+
+	// Multiple stage notifications
+	// - loaded
+	// - removed old shuttle
+	// - moved new shuttle to old location
+
+	//world.log << json_encode(data)
```
Comment (neersighted) [inline code/modules/shuttle/manipulator.dm]: delete :p

File: code/modules/shuttle/manipulator.dm [hunk 4]
```diff
@@ -9,5 +9,214 @@
...
+
+	var/mob/user = usr
+
+	switch(action)
+		if("select_template")
+			var/shuttle_id = params["shuttle_id"]
+
+			var/datum/map_template/shuttle/S = shuttle_templates[shuttle_id]
+			if(S)
+				src.selected = S
```
Comment (neersighted) [inline code/modules/shuttle/manipulator.dm]: Don't use src

File: code/modules/shuttle/manipulator.dm [hunk 5]
```diff
@@ -9,5 +9,214 @@
...
+				. = TRUE
+		if("jump_to")
+			if(params["type"] == "mobile")
+				for(var/i in SSshuttle.mobile)
+					var/obj/docking_port/mobile/M = i
+					if(M.id == params["id"])
+						user.loc = get_turf(M)
```
Comment (neersighted) [inline code/modules/shuttle/manipulator.dm]: I think we use forceMove instead of setting loc now

File: code/modules/shuttle/manipulator.dm [hunk 6]
```diff
@@ -9,5 +9,214 @@
...
+		if("preview")
+			var/shuttle_id = params["id"]
+			var/datum/map_template/shuttle/S = shuttle_templates[shuttle_id]
+			if(S)
+				load_template(S)
+				if(preview_shuttle)
+					preview_shuttle_id = shuttle_id
+					user.loc = get_turf(preview_shuttle)
```
Comment (neersighted) [inline code/modules/shuttle/manipulator.dm]: Ditto

File: code/modules/shuttle/manipulator.dm [hunk 7]
```diff
@@ -9,5 +9,214 @@
...
+		if("load")
+			var/shuttle_id = params["id"]
+			var/datum/map_template/shuttle/S = shuttle_templates[shuttle_id]
+			if(S)
+				if(!preview_shuttle || (S.shuttle_id != preview_shuttle_id))
```
Comment (neersighted) [inline code/modules/shuttle/manipulator.dm]: Tidy up this control flow?

File: code/modules/shuttle/manipulator.dm [hunk 8]
```diff
@@ -9,5 +9,214 @@
...
+					selected = null
+					return
+
+				// Destroy the old shuttle
+				existing_shuttle.jumpToNullSpace()
+				// Unflub the ID
+				var/new_id = replacetext(preview_shuttle.id, "(preview)", "")
+				preview_shuttle.id = new_id
+
+				preview_shuttle.dock(D)
```
Comment (neersighted) [inline code/modules/shuttle/manipulator.dm]: More comments would be good

File: code/modules/shuttle/manipulator.dm [hunk 9]
```diff
@@ -9,5 +9,214 @@
...
+
+				preview_shuttle = null
+				preview_shuttle_id = null
+				existing_shuttle = null
+				selected = null
+
+	update_icon()
+
+
+/obj/machinery/shuttle_manipulator/proc/load_template(
```
Comment (neersighted) [inline code/modules/shuttle/manipulator.dm]: Remove linebreak

File: code/modules/shuttle/manipulator.dm [hunk 10]
```diff
@@ -9,5 +9,214 @@
...
+				preview_shuttle.timer = timer
+				preview_shuttle.mode = mode
+				. = TRUE
+
+				preview_shuttle = null
+				preview_shuttle_id = null
+				existing_shuttle = null
+				selected = null
+
+	update_icon()
```
Comment (neersighted) [inline code/modules/shuttle/manipulator.dm]: Pretty sure there is no icon? maybe only update if `. = TRUE` (ie UI needs updating)?

File: code/modules/tgui/states.dm [hunk 11]
```diff
@@ -16,14 +16,20 @@
  **/
 /datum/proc/ui_status(mob/user, datum/ui_state/state)
 	var/src_object = ui_host()
+	. = UI_CLOSE
```
Comment (neersighted) [inline code/modules/tgui/states.dm]: I'm guessing the purpose of this is to allow `admin_state` when dead? Fair enough, that was an oversight when I added the ghost interaction.

File: run.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python
```
Comment (neersighted) [inline run.py]: What does this have to do with the PR?

File: tgui/assets/tgui.css [hunk 13]
```diff
@@ -1 +1,924 @@
...
\ No newline at end of file
+@charset "utf-8";
+html,
```
Comment (neersighted) [inline tgui/assets/tgui.css]: Assets need to be compiled with `--min`.

File: tgui/src/interfaces/shuttle_manipulator/templates.ract [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+			{{#each templates}}
+				<ui-display title='{{name}}'>
+					{{#if description}}
+						<ui-section label='Description'>{{description}}</ui-section>
+					{{/if}}
+					{{#if admin_notes}}
+						<ui-section label='Admin Notes'>{{admin_notes}}</ui-section>
+					{{/if}}
+
+<ui-button action='select_template'
```
Comment (neersighted) [inline tgui/src/interfaces/shuttle_manipulator/templates.ract]: Clean the indentation please (spaces, two or four. we like two mostly)

File: tgui/src/interfaces/shuttle_manipulator/status.ract [hunk 15]
```diff
@@ -0,0 +1,11 @@
+	{{#each data.shuttles}}
+		<ui-section label='{{name}} ({{id}})'>
+			{{status}}
+			{{#if timer}}
+				({{timeleft}})
+			{{/if}}
+<ui-button action='jump_to'
```
Comment (neersighted) [inline tgui/src/interfaces/shuttle_manipulator/status.ract]: Clean the indentation

File: tgui/src/interfaces/shuttle_manipulator/modification.ract [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+	{{/if}}
+	<ui-button action='jump_to'
+		params='{"type": "mobile", "id": "{{data.existing_shuttle.id}}"}'>
+	Jump To</ui-button>
+</ui-display>
+
+{{/if}}
+{{#if !data.existing_shuttle}}
+<ui-display title='Existing Shuttle: None'></ui-display>
+{{/if}}
```
Comment (neersighted) [inline tgui/src/interfaces/shuttle_manipulator/modification.ract]: Indentation helps readability

File: tgui/src/interfaces/shuttle_manipulator.ract [hunk 17]
```diff
@@ -0,0 +1,20 @@
+<link rel='ractive' href='./shuttle_manipulator/status.ract'>
+<link rel='ractive' href='./shuttle_manipulator/templates.ract'>
+<link rel='ractive' href='./shuttle_manipulator/modification.ract'>
+
+<ui-tabs tabs='{{data.tabs}}'>
+	<tab name='Status'>
+		<status></status>
```
Comment (neersighted) [inline tgui/src/interfaces/shuttle_manipulator.ract]: `<status/>`

File: tgui/src/interfaces/shuttle_manipulator.ract [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+	</tab>
+	<tab name='Templates'>
+		<templates></templates>
```
Comment (neersighted) [inline tgui/src/interfaces/shuttle_manipulator.ract]: ditto

File: tgui/src/interfaces/shuttle_manipulator.ract [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+	</tab>
+	<tab name='Modification'>
+		{{#if data.selected}}
+		<modification></modification>
```
Comment (neersighted) [inline tgui/src/interfaces/shuttle_manipulator.ract]: ditto

File: tgui/src/interfaces/shuttle_manipulator.ract [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+		{{/if}}
+		{{#if !data.selected}}
+		No shuttle selected.
```
Comment (neersighted) [inline tgui/src/interfaces/shuttle_manipulator.ract]: In a `<span class='bad'>` please?

File: tgui/build.sh [hunk 21]
```diff
@@ -0,0 +1,11 @@
+#RUN THIS IN THE tgui/ folder
+#!/bin/bash
```
Comment (neersighted) [inline tgui/build.sh]: Should be the first line on older kernels

File: tgui/build.sh [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+set -e
+export NODE_VERSION=4
+
+if [ ! -d "/tmp/nvm" ]; then
+	rm -rf /tmp/nvm && git clone https://github.com/creationix/nvm.git /tmp/nvm && (cd ~/.nvm && git checkout `git describe --abbrev=0 --tags`) && source /tmp/nvm/nvm.sh && nvm install $NODE_VERSION
```
Comment (neersighted) [inline tgui/build.sh]: Not thrilled with this script but oh well, supporting linux development is a good idea

File: tgui/build.sh [hunk 23]
```diff
@@ -0,0 +1,11 @@
...
+fi
+
+source /tmp/nvm/nvm.sh
+gulp
```
Comment (neersighted) [inline tgui/build.sh]: Just don't forget about gulp watch and `--min` before commit

File: code/modules/shuttle/shuttle.dm [hunk 24]
```diff
@@ -18,18 +18,16 @@
...
 	var/dheight = 0	//position relative to covered area, parallel to dir
 
-	var/i_know_what_im_doing = 0
-
 	//these objects are indestructable
-/obj/docking_port/Destroy()
+/obj/docking_port/Destroy(force)
 	// unless you assert that you know what you're doing. Horrible things
 	// may result.
-	if(!i_know_what_im_doing)
```
Comment (neersighted) [inline code/modules/shuttle/shuttle.dm]: classy

File: code/modules/shuttle/shuttle.dm [hunk 25]
```diff
@@ -346,25 +346,13 @@
...
-		T0.ChangeTurf(turf_type)
-
-		T0.redraw_lighting()
-		SSair.remove_from_active(T0)
-		T0.CalculateAdjacentTurfs()
-		SSair.add_to_active(T0,1)
-
-	src.i_know_what_im_doing = TRUE
-	qdel(src)
+	qdel(src, force=TRUE)
```
Comment (neersighted) [inline code/modules/shuttle/shuttle.dm]: thank you for using named parameters

File: code/modules/shuttle/manipulator.dm [hunk 26]
```diff
@@ -9,5 +9,214 @@
...
+
+
+/obj/machinery/shuttle_manipulator/proc/load_template(
+	datum/map_template/shuttle/S)
+	// load shuttle template, centred at shuttle import landmark,
+	var/turf/landmark_turf = get_turf(locate("landmark*Shuttle Import"))
```
Comment (neersighted) [inline code/modules/shuttle/manipulator.dm]: magic strings aaaaaaa

**Comment by neersighted** [comment]
heyoooo

File: code/modules/shuttle/manipulator.dm (hunk 9 above)
Comment (coiax) [inline code/modules/shuttle/manipulator.dm]: muh 79 character lines

File: code/modules/tgui/states.dm (hunk 11 above)
Comment (coiax) [inline code/modules/tgui/states.dm]: pretty much. It was a weird bug until I figured it out.

File: code/modules/shuttle/manipulator.dm (hunk 26 above)
Comment (Iamgoofball) [inline code/modules/shuttle/manipulator.dm]: thats the landmark system we cant really do much

File: code/game/gamemodes/blob/blobs/blob_mobs.dm [hunk 27]
```diff
@@ -271,3 +275,7 @@
 		factory.naut = null //remove this naut from its factory
 		factory.maxhealth = initial(factory.maxhealth)
 	flick("blobbernaut_death", src)
+
+/mob/living/simple_animal/hostile/blob/blobbernaut/independent
```
Comment (Shadowlight213) [inline code/game/gamemodes/blob/blobs/blob_mobs.dm]: How is this related to shuttles? Or am i missing something here?

Comment (coiax) [inline code/game/gamemodes/blob/blobs/blob_mobs.dm]: It has nothing to do with shuttles, but I discovered that a bolt of changed blobbernaut would start dying while making and testing the wabbajack shuttle.

Comment (Shadowlight213) [inline code/game/gamemodes/blob/blobs/blob_mobs.dm]: Atomize your prs meme

Comment (Iamgoofball) [inline code/game/gamemodes/blob/blobs/blob_mobs.dm]: nope, this would be a bugfix for the wabbajack

Comment (ChangelingRain) [inline code/game/gamemodes/blob/blobs/blob_mobs.dm]: Atomize the prs! Shove this into the polymorph pr you already have open.

File: code/modules/mob/mob_helpers.dm [hunk 28]
```diff
@@ -272,6 +272,10 @@ It's fairly easy to fix if dealing with single letters but not so much with comp
 			return M
 	return 0
 
+/proc/random_name()
```
Comment (RemieRichards) [inline code/modules/mob/mob_helpers.dm]: we already have one of these.

File: code/modules/shuttle/manipulator.dm [hunk 29]
```diff
@@ -6,8 +6,267 @@
...
+	datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, \
+	datum/ui_state/state = admin_state)
+
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "shuttle_manipulator", name, 800, 600, \
+			master_ui, state)
+		ui.open()
+
+/proc/shuttlemode2str(mode)
```
Comment (RemieRichards) [inline code/modules/shuttle/manipulator.dm]: this shouldn't exist, make the defines themselves strings.

File: code/modules/shuttle/emergency.dm [hunk 30]
```diff
@@ -6,61 +11,164 @@
...
+		ui.open()
+
+/obj/machinery/computer/emergency_shuttle/ui_data()
+	var/list/data = list()
+
+	data["timer_str"] = SSshuttle.emergency.getTimerStr()
+	data["engines_started"] = ENGINES_STARTED
+	data["authorizations_remaining"] = max((auth_need - authorized.len), 0)
+	var/A = list()
+	data["authorizations"] = A
```
Comment (RemieRichards) [inline code/modules/shuttle/emergency.dm]: this weirds me out so much.
I know lists are objects.
I know they're pass by reference.
but you're setting data before it's been added and it really just irks me.

File: code/modules/shuttle/emergency.dm [hunk 31]
```diff
@@ -6,61 +11,164 @@
...
+			400, 400, master_ui, state)
+		ui.open()
+
+/obj/machinery/computer/emergency_shuttle/ui_data()
+	var/list/data = list()
+
+	data["timer_str"] = SSshuttle.emergency.getTimerStr()
+	data["engines_started"] = ENGINES_STARTED
+	data["authorizations_remaining"] = max((auth_need - authorized.len), 0)
+	var/A = list()
```
Comment (RemieRichards) [inline code/modules/shuttle/emergency.dm]: this is a list, so typecast it.

File: code/game/objects/structures/tables_racks.dm [hunk 32]
```diff
@@ -325,18 +325,46 @@
...
+		// No climbing on the bar please
+		var/mob/living/M = AM
+		var/throwtarget = get_edge_target_turf(src, boot_dir)
+		M.Weaken(2)
+		M.throw_at_fast(throwtarget, 5, 1,src)
+		M << "<span class='notice'>No climbing on the bar please.</span>"
+	else
+		. = ..()
+
+/obj/structure/table/wood/bar/proc/is_barstaff(mob/user)
```
Comment (RemieRichards) [inline code/game/objects/structures/tables_racks.dm]: as said in IRC, check for bar access on the ID instead of this istype stuff.

**Comment by Bluespace** [comment]
Fun fact, this PR made rounds end instantly!

**Comment by Dorsisdwarf** [comment]
Nice
